### PR TITLE
planner: fix incorrect maintenance of `handleColHelper` for recursive CTE

### LINKED ISSUE #55666: nil pointer error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
``` sql
drop table if exists table_abc1;
drop table if exists table_abc2;
drop table if exists table_abc3;
drop table if exists table_abc4;

CREATE TABLE `table_abc1` (
  `column_abc1` varchar(10) DEFAULT NULL,
  `column_abc2` varchar(10) DEFAULT NULL,
  `column_abc3` varchar(10) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

CREATE TABLE `table_abc3` (
  `column_abc5` varchar(10) DEFAULT NULL,
  `column_abc6` varchar(10) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

CREATE TABLE `table_abc4` (
  `column_abc3` varchar(10) DEFAULT NULL,
  `column_abc7` varchar(10) DEFAULT NULL,
  `column_abc5` varchar(10) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

INSERT INTO `table_abc1` VALUES ('KTL157','KTL157','KTL157');
INSERT INTO `table_abc3` VALUES ('1000','20240819');
INSERT INTO `table_abc4` VALUES ('KTL157','test','1000');

DELETE FROM table_abc3 t_abc3
WHERE t_abc3.column_abc5 IN (
    SELECT a.column_abc5
    FROM (
        WITH tree_cte1 AS (
            WITH RECURSIVE tree_cte AS (
                SELECT t.column_abc1, t.column_abc2, t.column_abc3, 0 AS lv
                FROM table_abc1 t
                WHERE t.column_abc1 IN ('KTL157', 'KTL159')
                UNION ALL
                SELECT t.column_abc1, t.column_abc2, t.column_abc3, tcte.lv + 1
                FROM table_abc1 t
                JOIN tree_cte tcte ON t.column_abc1 = tcte.column_abc2
                WHERE tcte.lv <= 1
            )
            SELECT * FROM tree_cte
        )
        SELECT  e.column_abc5
        FROM (
            SELECT DISTINCT * FROM tree_cte1
        ) aa
        LEFT JOIN table_abc4 e ON e.column_abc3 = aa.column_abc3
    ) a
);


```

### 2. What did you expect to see? (Required)
no error

### 3. What did you see instead (Required)
```
ERROR 1105 (HY000): runtime error: invalid memory address or nil pointer dereference
```

```
[2024/08/26 18:36:20.488 +08:00] [INFO] [conn.go:1183] ["command dispatched failed"] [conn=872415234] [session_alias=] [connInfo="id:872415234, addr:127.0.0.1:56448 status:10, collation:utf8mb4_0900_ai_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="DELETE FROM table_abc3 t_abc3 WHERE t_abc3.column_abc5 IN (     SELECT a.column_abc5     FROM (         WITH tree_cte1 AS (             WITH RECURSIVE tree_cte AS (                 SELECT t.column_abc1, t.column_abc2, t.column_abc3, 0 AS lv                 FROM table_abc1 t                 WHERE t.column_abc1 IN ('KTL157', 'KTL159')                 UNION ALL                 SELECT t.column_abc1, t.column_abc2, t.column_abc3, tcte.lv + 1                 FROM table_abc1 t                 JOIN tree_cte tcte ON t.column_abc1 = tcte.column_abc2                 WHERE tcte.lv <= 1             )             SELECT * FROM tree_cte         )         SELECT  e.column_abc5         FROM (             SELECT DISTINCT * FROM tree_cte1         ) aa         LEFT JOIN table_abc4 e ON e.column_abc3 = aa.column_abc3     ) a )"] [txn_mode=PESSIMISTIC] [timestamp=452111520359514125] [err="runtime error: invalid memory address or nil pointer dereference\ngithub.com/pingcap/errors.AddStack\n\t/home/xhy/Development/Golang/GOPATH/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/errors.go:178\ngithub.com/pingcap/errors.Trace\n\t/home/xhy/Development/Golang/GOPATH/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20240318064555-6bd07397691f/juju_adaptor.go:15\ngithub.com/pingcap/tidb/pkg/util.GetRecoverError\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/util/util.go:288\ngithub.com/pingcap/tidb/pkg/executor/internal/exec.Next.func1\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/internal/exec/executor.go:435\nruntime.gopanic\n\t/home/xhy/.gvm/gos/go1.21.5/src/runtime/panic.go:914\nruntime.panicmem\n\t/home/xhy/.gvm/gos/go1.21.5/src/runtime/panic.go:261\nruntime.sigpanic\n\t/home/xhy/.gvm/gos/go1.21.5/src/runtime/signal_unix.go:861\ngithub.com/pingcap/tidb/pkg/executor.(*DeleteExec).deleteOneRow\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/delete.go:70\ngithub.com/pingcap/tidb/pkg/executor.(*DeleteExec).deleteSingleTableByChunk\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/delete.go:140\ngithub.com/pingcap/tidb/pkg/executor.(*DeleteExec).Next\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/delete.go:62\ngithub.com/pingcap/tidb/pkg/executor/internal/exec.Next\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/internal/exec/executor.go:451\ngithub.com/pingcap/tidb/pkg/executor.(*ExecStmt).next\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/adapter.go:1252\ngithub.com/pingc
ap/tidb/pkg/executor.(*ExecStmt).handleNoDelayExecutor\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/adapter.go:1001\ngithub.com/pingcap/tidb/pkg/executor.(*
ExecStmt).handleNoDelay\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/executor/adapter.go:834\ngithub.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec\n\t/home/xhy/Deve
lopment/github.com/pingcap/tidb/pkg/executor/adapter.go:598\ngithub.com/pingcap/tidb/pkg/session.runStmt\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/session/session
.go:2284\ngithub.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/session/session.go:2145\ngithub.com/pingcap/tidb/pk
g/server.(*TiDBContext).ExecuteStmt\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/server/driver_tidb.go:291\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).handleSt
mt\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/server/conn.go:2056\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).handleQuery\n\t/home/xhy/Development/github.com
/pingcap/tidb/pkg/server/conn.go:1810\ngithub.com/pingcap/tidb/pkg/server.(*clientConn).dispatch\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/server/conn.go:1384\ngi
thub.com/pingcap/tidb/pkg/server.(*clientConn).Run\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/server/conn.go:1146\ngithub.com/pingcap/tidb/pkg/server.(*Server).onC
onn\n\t/home/xhy/Development/github.com/pingcap/tidb/pkg/server/server.go:741\nruntime.goexit\n\t/home/xhy/.gvm/gos/go1.21.5/src/runtime/asm_amd64.s:1650"]
[2024/08/26 18:36:37.383 +08:00] [INFO] [domain.go:3156] ["instance plan eviction"] [detail="memory usage is below the soft limit, currentTot: 0, softLimit: 104857600"] [num
_evicted=0] [time_spent=850ns]

```

### 4. What is your TiDB version? (Required)

nightly

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #55666

Problem Summary:

If the recursive part of a CTE references tables other than the CTE itself, the `b.handleHelper` won't be correctly cleared during `buildRecursiveCTE()`.

It's because this function relies on an error from `buildSelect()` or `buildSetOpr()` to recognize the boundary between the seed part and the recursive part. But `b.handleHelper` won't be cleared when an error happens.

When this happens, there will be some wrong and redundant entries in `b.handleHelper`, which makes users of this field like `buildDelete()` get the wrong result or panic.

### What changed and how does it work?

Since this PR needs to be cherry-picked, I can't change too many.

- In `buildRecursiveCTE()`, clear the `b.handleHelper` if an error is returned from `buildSelect()` or `buildSetOpr()`.
- Try to make the `b.handleHelper` more clear for recursive CTE:
  - Now `tryBuildCTE()` will always push an entry into handleHelper when successful.
  - Now `buildRecursiveCTE()` will always push an entry into handleHelper when successful.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
Please add a description of how to handle CTE in the comments for the handleHelper attribute.

File: pkg/planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -7000,14 +7030,11 @@ func (b *PlanBuilder) buildRecursiveCTE(ctx context.Context, cte ast.ResultSetNo
 					// Build seed part plan.
 					saveSelect := x.SelectList.Selects
 					x.SelectList.Selects = x.SelectList.Selects[:i]
-					// We're rebuilding the seed part, so we pop the result we built previously.
```
Comment (Rustin170506) [inline pkg/planner/core/logical_plan_builder.go:7003]: Can I ask why we need to remove this?

**Review by Rustin170506 — APPROVED** [review]
Looks good to me. Thanks! 👍

File: pkg/planner/core/logical_plan_builder.go (hunk 1 above)
Comment (time-and-fate) [reply pkg/planner/core/logical_plan_builder.go:7003]: The `popMap()` here is moved right after `b.buildSelect()` and `b.buildSetOpr()` in this function.
This should make it more clear and explicit that we directly drop the entries pushed by them.

**Review by winoros — APPROVED** [review]

**Pull request merged**